src/Polly.Strategies/Retry/RetryStrategy.cs

```diff
@@ -0,0 +1,96 @@
+using Polly.Internals;
+
+namespace Polly.Retry;
+
+internal class RetryStrategy : DelegatingResilienceStrategy
+{
+    private readonly RetryStrategyOptions _options;
+    private readonly PredicatesHandler<ShouldRetryArguments> _shouldRetry;
+    private readonly EventsHandler<RetryActionArguments> _onRetry;
+    private readonly int _maxRetries;
+
+    public RetryStrategy(RetryStrategyOptions options)
+    {
+        _options = options;
+        _shouldRetry = PredicatesHandler<ShouldRetryArguments>.Create(options.ShouldRetry);
+        _onRetry = EventsHandler<RetryActionArguments>.Create(options.OnRetry);
+
+        _maxRetries = options.RetryCount;
+    }
+
+    public async override ValueTask<T> ExecuteAsync<T, TState>(Func<ResilienceContext, TState, ValueTask<T>> execution, ResilienceContext context, TState state)
+    {
+        var tryCount = 0;
+
+        while (true)
+        {
+            context.CancellationToken.ThrowIfCancellationRequested();
+
+            bool canRetry;
+            Outcome<T> outcome;
+
+            try
+            {
+                var result = await base.ExecuteAsync(execution, context, state).ConfigureAwait(context.ContinueOnCapturedContext);
+                outcome = new Outcome<T>(result);
+
+                if (!await _shouldRetry.HandleAsync(outcome, new ShouldRetryArguments(tryCount, context), context).ConfigureAwait(false))
+                {
+                    return result;
+                }
+
+                canRetry = tryCount < _maxRetries;
+                if (!canRetry)
+                {
+                    return result;
+                }
+
+                return result;
+            }
+            catch (Exception e)
+            {
+                if (!await _shouldRetry.HandleAsync(new Outcome<T>(e), new ShouldRetryArguments(tryCount, context), context).ConfigureAwait(context.ContinueOnCapturedContext))
+                {
+                    throw;
+                }
+
+                canRetry = tryCount < _maxRetries;
+                if (!canRetry)
+                {
+                    throw;
+                }
+
+                outcome = new Outcome<T>(e);
+            }
+
+            if (tryCount < int.MaxValue)
+            {
+                tryCount++;
+            }
+
+            var waitDuration = _options.RetryDelayGenerator(tryCount);
+
+            await _onRetry.HandleAsync(outcome, new RetryActionArguments(tryCount, context, waitDuration), context).ConfigureAwait(false);
+
+            await SleepAsync(context, waitDuration).ConfigureAwait(context.ContinueOnCapturedContext);
+        }
+    }
+
+    internal static Task SleepAsync(ResilienceContext context, TimeSpan delay)
+    {
+        if (delay == TimeSpan.Zero)
+        {
+            return Task.CompletedTask;
+        }
+
+        if (context.IsSynchronous)
+        {
+            Thread.Sleep(delay);
```

Comment:
We must virtualize all timer usage to enable robust testing.

Reply:
Yeah definitely, have not included it here for simplicity. I think we will have some property like `ResilienceStrategyBuilderOptions.SystemClock`. Similar to what's in the Polly now, but it won't be static anymore.

Reply:
That would be good - that would also fit with things like the [`ISystemClock`](https://github.com/dotnet/aspnetcore/blob/main/src/Middleware/OutputCaching/src/ISystemClock.cs) pattern used in various places inside ASP.NET Core.

Reply:
So, are you OK if I keep this out of POC or should I add the abstraction?

Reply:
I think it's fine to leave out for now.